### DISCUSSION TITLE
Saving of links to LinkAce on iOS devices

### DISCUSSION THREAD

**Discussion by Kovah** [Kovah/LinkAce#364] [Guides]
At the moment LinkAce has no mobile apps, so adding links on your mobile phone can be tedious. But LinkAce provides a great API that allows you to access and edit all data from anywhere. With the help of the API and Shortcuts, iOS users can easily add links to LinkAce from the share menu. I created a Shortcuts workflow that uses only two inputs from you: your LinkAce URL and your API key which can be found in the user settings.

## Install the iOS Shortcut

By default, iOS Shortcuts does not allow the installation of “untrusted” Shortcuts shared via URL. It will deny the installation of shortcuts because “Shortcuts security settings don’t allow untrusted shortcuts.,” Before you install the Shortcut, enable untrusted sources:

* go to your iOS settings
* go to the Shortcuts app settings
* enable “Allow untrusted Shortcuts”

There are two different versions of the shortcut available. Click on either link in Safari on your phone and install the Shortcut.

**v2.x**
* [**Shortcut asking for optional title**](https://www.icloud.com/shortcuts/46b654894c53497d8f86d9a1a9b6879e)
* [**Shortcut asking for title, description, lists and tags**](https://www.icloud.com/shortcuts/1505d128cd204bdb9a29293c6f297edf)

**v1.x**
* [**Shortcut asking for optional title**](https://www.icloud.com/shortcuts/b8741345ad794d158bbf1b91a4d515bf)
* [**Shortcut asking for title, description, lists and tags**](https://www.icloud.com/shortcuts/d0c6b9bbb74a4f75aa8b7a3c0ff4278c)

You will now be redirected to the Shortcuts app with the workflow preview. It contains many elements needed to successfully save links in LinkAce. There are two things you must edit while importing the Shortcut: the LinkAce URL and your user API key. Click the “Import untrusted Shortcut” now. The app will ask you for both the URL and the key in the next two steps. After that, the shortcut is ready to use.

## Details about the Shortcut workflow

* First, the Shortcut is enabled for the share menu and will only accept URLs as the primary input source. If you want to use other input sources, like the clipboard, you should tweak the workflow.
* In the first steps, the primary input, the URL and the API key are saved to variables. After clicking the Shortcut in the share menu, it will ask for input: you can either enter a custom link title now, or leave it blank to let LinkAce generate it for you.
* After that, a network request is sent to your LinkAce URL including the correct API path: /api/v1/links, for example http://demo.linkace.org/api/v1/links. This network request contains the shared URL, the title if provided, and some standard values for other fields.
* To make the Shortcut a bit nicer, the response of the API is parsed in the last steps. If the response contains a link ID, the shortcut assumes that the link was saved successfully. If not, most likely an error was returned. If an error was returned, it will be shown to you.

---

_This tutorial was first published on [blog.kovah.de](https://blog.kovah.de/en/2020/ios-shortcuts-workflow-for-linkace/)_

**Comment by craigmckay** [comment]
Fantastic Kovah!  Funktioniert ein Genuss!

**Comment by craigmckay** [comment]
I'm going to amend the Shortcut to remove the OK/Cancel "Going to post now" (is there a reason for that one) and then the final one to "Open in browser" - just to make as quick as possible.

**Comment by smgt** [comment]
Just wanted to add my take on the iOS Shortcut. I have two shortcuts, one "Bookmark" and one "Bookmark Fast". The "Bookmark" Shortcut uses the same url as the bookmarklet and opens it in "Show web page at" step. There you can fill the form as you see fit.
The "Bookmark Fast" Shortcut just sends the url to the linkace api for adding links without any input from the user.

**Reply by ckykenken** [reply]
How can I achieve this? I am using Edge browser and it has been blocking JavaScripts a few months ago so I have to use shortcuts

**Comment by kpleines** [comment]
Sorry for my question, but I can't find a documentation of that feature for android devices. Is ther any documentation?
thx for help

**Reply by Kovah** [reply]
Hey, I am pretty sure there is no similar feature in Android yet. Other users suggested having such a feature in LinkAce that can be "installed" on smartphone, but it currently has no priority for me personally.

**Reply by sergiorgiraldo** [reply]
@kpleines 

For Android, you can implement on your site if you just want a fast way to add a link:
1. Create a `manifest.json` file on root folder like this:
```
{
    "short_name": "Share to Linkace",
    "name": "Share to Linkace",
    "share_target": {
        "action": "public/bookmarklet/add",
        "method": "GET",
        "params": {
            "title": "t",
            "text": "u"
        }
    },
    "description": "Share from mobile to linkace",
    "icons": [
        {
            "src": "/public/android-chrome-192x192.png",
            "sizes": "192x192",
            "type": "image/png"
        },
        {
            "src": "/public/android-chrome-384x384.png",
            "sizes": "512X512",
            "type": "image/png"
        }
    ],
    "start_url": "/",
    "background_color": "#f3e5f5",
    "display": "standalone",
    "scope": "/",
    "theme_color": "#4a148c"
}
```
2. Add this to the head:
`<link href="/manifest.json" rel="manifest" />`
3. Now open the homepage in chrome in your android (dont use other browser), got the chevron and select "Install App"
4. After installed you can see linkace as an option when you share something. Beware not all apps play nicely and expose the title and url, for those, you have to manually input the fields (chrome and edge are my usual sources, they are ok with this hack)

**Comment by barrymieny** [comment]
I've just added the shortcuts and had to make some changes to make them work, but apart from that they work great.

Could you update the shortcuts shared here to make use of `/api/v2/links` and the `authorization: Bearer ...` header?
